autonat/autonat-v2.md

```diff
@@ -0,0 +1,207 @@
+the message.
+
+`E_INTERNAL_ERROR`: error not classified within the above error codes occured on
+peer that prevented it from completing the request.
+
+Implementations MUST discard responses with status codes they do not understand
+
+### Consideration for DDOS Prevention
+
+In order to prevent attacks like the one described in [RFC 3489, Section
+12.1.1](https://www.rfc-editor.org/rfc/rfc3489#section-12.1.1) (see excerpt
+below), implementations MUST NOT dial any multiaddress unless it is based on the
+IP address the requesting node is observed as. This restriction as well implies
+that implementations MUST NOT accept dial requests via relayed connections as
+one can not validate the IP address of the requesting node.
+
+> RFC 3489 12.1.1 Attack I: DDOS Against a Target
+>
+> In this case, the attacker provides a large number of clients with the same
+> faked MAPPED-ADDRESS that points to the intended target. This will trick all
+> the STUN clients into thinking that their addresses are equal to that of the
+> target. The clients then hand out that address in order to receive traffic on
+> it (for example, in SIP or H.323 messages). However, all of that traffic
+> becomes focused at the intended target. The attack can provide substantial
+> amplification, especially when used with clients that are using STUN to enable
+> multimedia applications.
+
+
+## Implementation Suggestions
+
+For any given address, implementations SHOULD do the following
+- periodically recheck reachability status
+- query multiple peers to determine reachability
+
+The suggested heuristic for implementations is to consider an address reachable
+if more than 3 peers report a successful dial and to consider an address
+unreachable if more than 3 peers report unsuccessful dials. 
+
+Implementations are free to use different heuristics than this one
+
+
+## RPC Messages
+
+All RPC messages sent over a stream are prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec]. 
+
+All RPC messages on stream `/libp2p/autonat/2.0.0/dial` are of type
+`DialMessage`. A `DialRequest` message is sent as a `DialMessage` with the `dialRequest`
+field set and the `type` field set to `DIAL_REQUEST`. `DialResponse` is handled
+similarly.
+
+On stream `/libp2p/autonat/2.0.0/attempt`, there is a single message type
+`AttemptMessage`
+
+```proto
+syntax = "proto3";
+
+message DialMessage {
+  enum Type {
+    DIAL_REQUEST  = 0;
+    DIAL_RESPONSE = 1;
+  }
+
+  Type type          = 1;
+  DialRequest dialRequest   = 2;
+  DialResponse dialResponse = 3;
+}
+
+message Candidate {
+  bytes addr = 1;
+  fixed64 nonce = 2;
+}
+
+message DialRequest {
+  repeated Candidate candidates = 1;
+}
+
+message DialResponse {
+  enum ResponseStatus {
+      OK                        = 0;
+      E_DIAL_ERROR              = 100;
+      E_DIAL_REFUSED            = 101;
```

Comment:
What is the difference between error and refused? What are examples for each?

Reply:
in the section Requirements for ResponseStatus there are these lines 
```
E_DIAL_ERROR`: the peer attempted to dial _addr_ and was unable to connect. 

`E_DIAL_REFUSED`: the peer didn't attempt a dial because of rate limiting,
resource limit reached or blacklisting.
```
Is this not clear enough? Should I add an example? 
https://github.com/libp2p/specs/blob/0771babb17eb9e33d5334c78cb658fdf7e687a7f/autonat/autonat-v2.md?plain=1#L90-L94

Reply:
Please ignore my comment. I didn't read attentively.